### PR TITLE
[py-torchvision] added dependency of ffmpeg

### PR DESCRIPTION
**Pull request by qwertos**

### CODE REVIEW

**Comment by adamjstewart** [comment]
Can you add more details as to the motivation of this PR? Where do you see this dependency? Is it only required for certain datasets?

**Comment by qwertos** [comment]
Solving this build issue:

```
==> 31427: Installing py-torchvision
==> Using cached archive: /home/jchits/workspace/spack/var/spack/cache/_source-cache/archive/02/02de11b3abe6882de4032ce86dab9c7794cbc84369b44d04e667486580f0f1f7.tar.gz
==> Staging archive: /dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/v0.6.0.tar.gz
==> Created stage in /dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp
==> No patches needed for py-torchvision
==> 31427: py-torchvision: Building py-torchvision [PythonPackage]
==> 31427: py-torchvision: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 1:
    '/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/python-3.8.3-2fkhskcjshkm3q6gz6kmrbaob7ge5zma/bin/python3.8' '-s' 'setup.py' '--no-user-cfg' 'build'

7 errors found in build log:
     603     #define CHECK_EQ(val1, val2) CHECK_OP(val1, val2, ==)
     604                                  ^~~~~~~~
     605    /dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/torchvision/csrc/cpu/video_reader/VideoReader.cpp:95:3: note: in expansion of macro 'CHECK_EQ'
     606       CHECK_EQ(framePts.size(0), msgs.size());
     607       ^~~~~~~~
     608    /home/jchits/workspace/spack/lib/spack/env/gcc/g++ -pthread -shared -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/sqlite-3.31.1-izknovbz3ud2lhl5elmtdjxyp3nlqvkf/lib -L/ho
            me/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/zlib-1.2.11-2famardgbouq6bbvoj2tn43h2h7xgrhj/lib -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/bz
            ip2-1.0.8-xnavnah75h3ytapkvfczxnzdod6ifdkr/lib -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/ncurses-6.2-onh7rphci3xzzwemi5jogzjjqwxcyagk/lib -L/home/jchits/workspace/spa
            ck/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/expat-2.2.9-4uijjrjqymvvfalfa34xf5iawkk2bjt6/lib -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/openssl-1.1.1g-cr54apn44l
            gvlypntby4u3pnolxrym6l/lib -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/gettext-0.20.2-coanxqpwxnicfkt7hshsr2a6iiivnkcj/lib -L/home/jchits/workspace/spack/opt/spack/linu
            x-rhel7-skylake_avx512/gcc-8.2.0/xz-5.2.5-pole6xvjm74ks4zqye5cjplyn2mj43dj/lib -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/tcl-8.6.8-7acxe4sam63suyzzgfpsiu57wxrqedti/li
            b -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/tk-8.6.8-bxeq5jbf5xxop6ckkljh6gxt7axjeblx/lib -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.
            0/libffi-3.3-u3phmoraw5yjyg6hj2ikszlyoxbmwpf3/lib64 -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/readline-8.0-ustvssxoiftx7m6tvdeynwt4ciq4dpjs/lib -L/home/jchits/workspa
            ce/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/gdbm-1.18.1-qy5adikdbzlh5vfdl4fqzbu2u5crdcwv/lib /dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spa
            ck-src/build/temp.linux-x86_64-3.8/dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/torchvision/csrc/cpu/video_reader/VideoReader.o /dev/shm/spack/s
            tage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/build/temp.linux-x86_64-3.8/dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmf
            quajkp/spack-src/torchvision/csrc/cpu/decoder/video_stream.o /dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/build/temp.linux-x86_64-3.8/dev/shm/s
            pack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/torchvision/csrc/cpu/decoder/video_sampler.o /dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m4
            6652gaeuz5douxkpfry4rdmfquajkp/spack-src/build/temp.linux-x86_64-3.8/dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/torchvision/csrc/cpu/decoder/m
            emory_buffer.o /dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/build/temp.linux-x86_64-3.8/dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0
            .6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/torchvision/csrc/cpu/decoder/audio_stream.o /dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/build/
            temp.linux-x86_64-3.8/dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-src/torchvision/csrc/cpu/decoder/audio_sampler.o -L/home/jchits/workspace/spack/o
            pt/spack/linux-rhel7-skylake_avx512/gcc-8.2.0/py-torch-1.5.0-hcezcfbakqwvq3f6vtbxen4n6zet6uvr/lib/python3.8/site-packages/torch/lib -L/home/jchits/workspace/spack/opt/spack/linux-rhel7-skylake_avx512/gc
            c-8.2.0/python-3.8.3-2fkhskcjshkm3q6gz6kmrbaob7ge5zma/lib -lavcodec -lavformat -lavutil -lswresample -lswscale -lc10 -ltorch -ltorch_cpu -ltorch_python -o build/lib.linux-x86_64-3.8/torchvision/video_re
            ader.so -std=c++14
  >> 609    /usr/bin/ld: cannot find -lavcodec
  >> 610    /usr/bin/ld: cannot find -lavformat
  >> 611    /usr/bin/ld: cannot find -lavutil
  >> 612    /usr/bin/ld: cannot find -lswresample
  >> 613    /usr/bin/ld: cannot find -lswscale
  >> 614    collect2: error: ld returned 1 exit status
  >> 615    error: command '/home/jchits/workspace/spack/lib/spack/env/gcc/g++' failed with exit status 1

See build log for details:
  /dev/shm/spack/stage/jchits/spack-stage-py-torchvision-0.6.0-m46652gaeuz5douxkpfry4rdmfquajkp/spack-build-out.txt
```

**Comment by adamjstewart** [comment]
Interesting, I don't have ffmpeg installed on my laptop, but the build works just fine. Do you have an `ffmpeg` executable in your `PATH`? Based on the `setup.py`, it looks like ffmpeg is used if the executable is found, but maybe your libraries are in a weird location and aren't being picked up. Without hacking the `setup.py`, I don't see an easy way to make the dependency optional.

Also, based on https://github.com/pytorch/vision/issues/1597, the dependency should be ffmpeg 3.1+.

**Comment by qwertos** [comment]
In this case, it's being concretized into a spack environment where ffmpeg is being brought in as a dependency of something else. This leads me to believe that it's finding ffmpeg in the view directory for the spack environment set in PATH, but none of the library paths are set.

File: var/spack/repos/builtin/packages/py-torchvision/package.py [hunk 1]
```diff
@@ -42,3 +42,5 @@ class PyTorchvision(PythonPackage):
     # Many of the datasets require additional dependencies to use.
     # These can be installed after the fact.
     depends_on('py-scipy', type='test')
+
+    depends_on('ffmpeg@3.1:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-torchvision/package.py]:
```suggestion
    depends_on('ffmpeg@3.1:', when='@0.4.2:')
```
https://github.com/pytorch/vision/releases/tag/v0.4.2

**Pull request merged**